### ISSUE TITLE
CursorPagination on a FloatField will result in duplicate elements on page boundary

### ISSUE THREAD

**Issue by keturn** [encode/django-rest-framework#5160]
## Checklist

- [ ] I have verified that that issue exists against the `master` branch of Django REST framework.
- [x] I have searched for similar issues in both open and closed tickets and cannot find a duplicate.
- [x] This is not a usage question. (Those should be directed to the [discussion group](https://groups.google.com/forum/#!forum/django-rest-framework) instead.)
- [x] This cannot be dealt with as a third party library. (We prefer new functionality to be [in the form of third party libraries](http://www.django-rest-framework.org/topics/third-party-resources/#about-third-party-packages) where possible.)
- [ ] I have reduced the issue to the simplest possible case.
- [ ] I have included a failing test as a pull request. (If you are unable to do so we can still accept the issue.)

## Steps to reproduce

Use CursorPagination on a ModelViewSet, ordering by a FloatField.

Prototype of a test here. (I'm not used to writing project-independent Django application tests, this will need a little tweaking before it runs.)
https://gist.github.com/keturn/8bc88525a183fd41c73ffb729b8865be

## Expected behavior

Expect ViewSet to provide pages that iterate through all the results with no duplicates.

## Actual behavior

After following the `next` cursor to the next page, that second page begins with the last item on the first page.

This is because the cursor's representation of the floating-point position has been truncated, so the last item falls on the wrong side of the inequality test.

**Comment by xordoquy** [comment]
Could you try with:

    queryset = FloatyModel.objects.all().order_by('-score')

as I'm not sure ordering will happen without an ordering filter.

**Comment by keturn** [comment]
We're working around it by limiting the precision on the cursor to the point where it doesn't lose anything on round-trip through psycopg. https://gist.github.com/keturn/8bc88525a183fd41c73ffb729b8865be#file-fpcursorpagination-py

**Comment by xordoquy** [comment]
I don't get your comment here.

**Comment by keturn** [comment]
More background on the phenomenon at https://psycopg.lighthouseapp.com/projects/62710/tickets/145-floats-are-not-exactly-preserved

this isn't really DRF's bug, it's a "psycopg/postgresql text representations of floating point numbers" bug, but since CursorPagination depends on ordering floating point fields, it's real easy to run in to this.

**Comment by cypreess** [comment]
@keturn I would highly advice you not to use `float` for anything that you rely on as a value (and in CursorPaginator you highly rely on this value !). Instead I would definitelly use Decimal field (simillary  to any financial applications) with specific decimal precision that meets your needs (you even claimed by yourself that you solved problem by cutting precision to some specific point that dont makes this problem appearing). By using Decimal you will just straightforwadly declare what is your precision that you work with, and will store it always in the same way. For example this is exactly how I would store float timestamps from python.

Seriously, I don't see any other soulution for that, as this is just a design flaw of general floats representation in computer systems.

**Comment by tomchristie** [comment]
@cypreess Agreed really. I think we should probably address this as a documentation bug.

**Comment by keturn** [comment]
I'd be sympathetic to this argument if the use case were something like "cursor is at `0.2` and user advances it by `0.1` and the entry at `0.3` is skipped," but that's not how DRF presents CursorPagination. The cursor position isn't calculated from user input — or indeed, calculated from *anything* in Python — that's the position is the value of a particular field in the result set, and the framework, not the user, is making all the decisions about how that value is represented and used in comparisons.

There's no uncertainty principle or fundamental problem of computer science that says this can't work.

I *am* sympathetic to the fact that it's something down near the postgresql-protocol/psycopg2 layer that's using a lossy string representation of that value, and that's well out of scope of DRF to solve directly.

Whether it's something we can work around in a Pagination class, or a thing we have to document how the library user should change their query, I would like that to be about changing the *query,* not the schema of the underlying table.

...oh wait, I just remembered that normal people just compose their queries with the Django ORM managed model classes instead of the crazy combination of database views and `annotate` clauses we're using in this project, but I believe the principle still stands.

**Comment by carltongibson** [comment]
From #5164: 

https://github.com/veeloinc/django-rest-framework/blob/e9df63a467bc939aa8db29bc95a128f440e5b517/rest_framework/pagination.py#L764-L769

This isn't a big change. (The `_rounding_up...`/`_rounding_down...` defs are a line each.) 

**BUT** I am inclined to think that using a `FloatField` here is kind of a _programming error_. (Why? Because `float`s. We **know** this is going to happen somewhere...) 

>  I think we should probably address this as a documentation bug.

Adding a "Should not be a float" to [this list of caveats in the docs](https://github.com/encode/django-rest-framework/blame/master/docs/api-guide/pagination.md#L179-L182) seems the _right way_ to go. (Sorry about the Blame view there, GitHub won't show me an un-rendered markdown file.) 

**BUT** that might be a bit mean: 

>  I would like that to be about changing the query, not the schema of the underlying table.

Maybe we already used a float and need to paginate on it. 

In that case I think linking to @keturn's [`FloatingPointCursorPagination` gist](https://gist.github.com/keturn/8bc88525a183fd41c73ffb729b8865be#file-fpcursorpagination-py) example, whilst saying "Don't do this", is a good compromise.

**Comment by amankumarjain** [comment]
Hi guys, with using cursor pagination it solved my duplication items on boundry once user add or delete in parallel, and my question is lets say on home page i got a cursor for next page like:
```
{
    "next": "http://{url}/api/v1/wgroup/?cursor=cD0yMDE4LTA1LTE3KzIxJTNBMjMlM0EzOS4xNDMxMTMlMkIwMCUzQTAw&q=12",
    "previous": null,
    "results": [
        {
            "id": "721d43b9-0034-47c1-9c1e-f06e9ed7a534",
            "title": "Redmi phone booking",
       }
]
```

if user click on next and in parallel another user added a post how i will get all new post count exist for before this cursor value. i need to show user like {5 new post available}
